File: 63. Unique Paths II/63. Unique Paths II top-down.py
from functools import lru_cache
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        @lru_cache(None)
        def dp(row: int, column: int) -> int:
            if obstacleGrid[row][column] == 1:
                return 0
            if row+column == 0:
                return 1
            ways = 0
            if row > 0:
                ways += dp(row-1, column)
            if column > 0:
                ways += dp(row, column-1)
            return ways
        return dp(m-1, n-1)

File: 63. Unique Paths II/63. Unique Paths II top-down.java
class Solution {
    private int[][] memo, obstacleGrid;

    private int dp(int row, int column) {
        if (obstacleGrid[row][column] == 1) {
            return 0;
        }
        if (row + column == 0)
            return 1;
        if (memo[row][column] == 0) {
            int ways = 0;
            if (row > 0) {
                ways += dp(row - 1, column);
            }
            if (column > 0) {
                ways += dp(row, column - 1);
            }
            memo[row][column] = ways;
        }
        return memo[row][column];
    }

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        this.obstacleGrid = obstacleGrid;
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        this.memo = new int[m][n];
        return dp(m - 1, n - 1);
    }
}